File: membrain-node/src/cbr/prompt_builder.ts
/**
 * Prompt builder for in-context learning with retrieved cases.
 *
 * Formats positive and negative cases into sections that can be prepended to
 * an LLM prompt so the model learns from past experience.
 */

import type { CaseEntry, CaseSearchResults } from "../types";

export interface CasePromptBuilderOptions {
  maxPositiveExamples?: number;
  maxNegativeExamples?: number;
}

export class CasePromptBuilder {
  private readonly maxPositive: number;
  private readonly maxNegative: number;

  constructor(options: CasePromptBuilderOptions = {}) {
    this.maxPositive = options.maxPositiveExamples ?? 3;
    this.maxNegative = options.maxNegativeExamples ?? 2;
  }

  buildContext(cases: CaseSearchResults): string {
    const sections: string[] = [];

    const positive = cases.positive_cases.slice(0, this.maxPositive);
    if (positive.length > 0) {
      sections.push(renderSection("Successful Past Approaches", "Example", positive));
    }

    const negative = cases.negative_cases.slice(0, this.maxNegative);
    if (negative.length > 0) {
      sections.push(
        renderSection("Approaches to Avoid", "Counter-Example", negative)
      );
    }

    return sections.join("\n");
  }
}

function renderSection(
  heading: string,
  itemLabel: string,
  cases: CaseEntry[]
): string {
  const lines: string[] = [`## ${heading}`, ""];
  cases.forEach((entry, index) => {
    lines.push(`### ${itemLabel} ${index + 1}`);
    lines.push(`**Problem:** ${entry.problem}`);
    lines.push(`**Plan:**\n${formatPlan(entry.plan)}`);
    lines.push(`**Outcome:** ${entry.outcome}`);
    lines.push("");
  });
  return lines.join("\n");
}

export function formatPlan(plan: string): string {
  let parsed: unknown;
  try {
    parsed = JSON.parse(plan);
  } catch {
    return plan;
  }

  if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
    const record = parsed as Record<string, unknown>;
    if ("plan" in record) {
      parsed = record.plan;
    }
  }

  if (!Array.isArray(parsed)) {
    return plan;
  }

  return parsed
    .map((step, index) => {
      if (step && typeof step === "object" && !Array.isArray(step)) {
        const description = (step as Record<string, unknown>).description;
        const text =
          typeof description === "string" ? description : JSON.stringify(step);
        return `${index + 1}. ${text}`;
      }
      return `${index + 1}. ${String(step)}`;
    })
    .join("\n");
}

File: membrain-py/membrain/cbr/prompt_builder.py
"""Prompt builder for in-context learning with retrieved cases.

Formats positive and negative cases into sections that can be prepended
to an LLM prompt so the model learns from past experiences.
"""

from __future__ import annotations

import json

from ..types import CaseSearchResults


class CasePromptBuilder:
    """Builds in-context learning prompts from retrieved cases.

    Positive cases become "Successful Past Approaches" and negative cases
    become "Approaches to Avoid".
    """

    def __init__(
        self,
        max_positive_examples: int = 3,
        max_negative_examples: int = 2,
    ) -> None:
        self._max_positive = max_positive_examples
        self._max_negative = max_negative_examples

    def build_context(self, cases: CaseSearchResults) -> str:
        """Build a prompt context string from retrieved cases.

        Args:
            cases: The positive and negative cases from a retriever.

        Returns:
            A formatted string to include as context in the LLM prompt.
            Returns an empty string if no cases are available.
        """
        sections: list[str] = []

        positive = cases.positive_cases[: self._max_positive]
        if positive:
            lines = ["## Successful Past Approaches", ""]
            for index, case in enumerate(positive, start=1):
                lines.append(f"### Example {index}")
                lines.append(f"**Problem:** {case.problem}")
                lines.append(f"**Plan:**\n{_format_plan(case.plan)}")
                lines.append(f"**Outcome:** {case.outcome}")
                lines.append("")
            sections.append("\n".join(lines))

        negative = cases.negative_cases[: self._max_negative]
        if negative:
            lines = ["## Approaches to Avoid", ""]
            for index, case in enumerate(negative, start=1):
                lines.append(f"### Counter-Example {index}")
                lines.append(f"**Problem:** {case.problem}")
                lines.append(f"**Plan:**\n{_format_plan(case.plan)}")
                lines.append(f"**Outcome:** {case.outcome}")
                lines.append("")
            sections.append("\n".join(lines))

        return "\n".join(sections)


def _format_plan(plan: str) -> str:
    """Format a plan string for display.

    If the plan is valid JSON with a list of steps, formats them as
    a numbered list. Otherwise returns the plan text as-is.
    """
    try:
        parsed = json.loads(plan)
    except (json.JSONDecodeError, TypeError):
        return plan

    if isinstance(parsed, dict) and "plan" in parsed:
        parsed = parsed["plan"]

    if isinstance(parsed, list):
        lines = []
        for index, step in enumerate(parsed, start=1):
            if isinstance(step, dict):
                description = step.get("description", str(step))
            else:
                description = str(step)
            lines.append(f"{index}. {description}")
        return "\n".join(lines)

    return plan
